test/cpp_api_parity/utils.py

```diff
@@ -0,0 +1,332 @@
+        # }
+        # ```
+        # )
+        'arg_dict',
+
+        # Whether we expect this NN functional test to pass the Python/C++ parity test
+        # (e.g. `True`)
+        'has_parity',
+
+        # Device (e.g. "cuda")
+        'device',
+
+        # Temporary folder to store C++ outputs (to be compared with Python outputs later)
+        'cpp_tmp_folder',
+    ]
+)
+
+CppArg = namedtuple('CppArg', ['name', 'value'])
+
+TORCH_NN_COMMON_TEST_HARNESS = """
+#include <torch/script.h>
+
+void write_ivalue_to_file(const torch::IValue& ivalue, const std::string& file_path) {
+    auto bytes = torch::jit::pickle_save(ivalue);
+    std::ofstream fout(file_path, std::ios::out | std::ios::binary);
+    fout.write(bytes.data(), bytes.size());
+    fout.close();
+}
+
+c10::Dict<std::string, torch::Tensor> load_dict_from_file(const std::string& file_path) {
+    c10::Dict<std::string, torch::Tensor> arg_dict;
+    auto arg_dict_module = torch::jit::load(file_path);
+    for (const auto& p : arg_dict_module.named_buffers(/*recurse=*/false)) {
+        arg_dict.insert(p.name, p.value);
+    }
+    return arg_dict;
+}
+
+// Generates rand tensor with non-equal values. This ensures that duplicate
+// values won't be causing test failure for modules like MaxPooling.
+// size should be small, otherwise randperm fails / long overflows.
+torch::Tensor _rand_tensor_non_equal(torch::IntArrayRef size) {
+    int64_t total = 1;
+    for (int64_t elem : size) {
+        total *= elem;
+    }
+    return torch::randperm(total).view(size).to(torch::kDouble);
+}
+"""
+
+def compile_cpp_code_inline(name, cpp_sources, functions):
+    cpp_module = torch.utils.cpp_extension.load_inline(
+        name=name,
+        cpp_sources=cpp_sources,
+        functions=functions,
+        verbose=False,
+    )
+    return cpp_module
+
+def compute_temp_file_path(cpp_tmp_folder, variant_name, file_suffix):
+    return os.path.join(cpp_tmp_folder, '{}_{}.pt'.format(variant_name, file_suffix))
+
+def is_torch_nn_functional_test(test_params_dict):
+    return 'wrap_functional' in str(test_params_dict.get('constructor', ''))
+
+def convert_to_list(python_input):
+    if isinstance(python_input, torch.Tensor):
+        return [python_input]
+    else:
+        return list(python_input)
+
+def set_python_tensors_requires_grad(python_tensors):
+    return [tensor.requires_grad_(True) if tensor.dtype != torch.long else tensor for tensor in python_tensors]
+
+def move_python_tensors_to_device(python_tensors, device):
+    return [tensor.to(device) for tensor in python_tensors]
+
+def has_test(unit_test_class, test_name):
+    return hasattr(unit_test_class, test_name)
+
+def add_test(unit_test_class, test_name, test_fn):
+    if has_test(unit_test_class, test_name):
+        raise RuntimeError("Found two tests with the same name: " + test_name)
+    setattr(unit_test_class, test_name, test_fn)
+
+def set_cpp_tensors_requires_grad(cpp_tensor_stmts, cpp_tensors):
```

Comment:
What exactly is `cpp_tensors`? From the zip statement below it doesn't look like a list of python `torch.Tensor`

Reply:
It actually is a list of python `torch.Tensor` - I renamed the argument to `python_tensors` to make it more clear